Add Helm option to disable registering CRD from Cilium Operator

Fedosin opened the pull request:
This PR contains three commits that: 
* add a Helm option `skipCRDCreation` for Cilium Operator that allows to skip CRDs creation;
* add `--skip-crd-creation` flag for the operator, indicating that CRDs creation should be skipped;
* deprecate `option.Config.SkipCRDCreation` and `--skip-crd-creation` for cilium-agent.

Fixes: #15262

Fedosin committed:
Allow to skip CRDs creation for the operator

This commit adds `--skip-crd-creation` flag for the operator,
indicating that CRDs creation should be skipped.

Signed-off-by: Mike Fedosin <mfedosin@gmail.com>

Fedosin committed:
Add helm option to disable registering CRD from Cilium Operator

This commit adds a Helm option skipCRDCreation for Cilium
Operator that allows to skip CRDs creation.

Signed-off-by: Mike Fedosin <mfedosin@gmail.com>

aanm commented on operator/main.go:
```diff
@@ -274,8 +274,12 @@ func runOperator() {
+			log.WithError(err).Fatal("Unable to register CRDs")
+		}
+	} else {
+		log.Info("Skipping creation of CRDs")
```
Hmm, maybe it's a good idea to move `operatorOption.Config.SkipCRDCreation` to inside the `RegisterCRDs` function and remove the other option is in there (`option.Config.SkipCRDCreation`). As well as remove the `SkipCRDCreation` variable from the option.Config altogether.

kaworu replied:
@aanm I wasn't able to parse your sentence, could you rephrase please?

aanm replied:
@kaworu I've edited the comment.

Fedosin replied:
@aanm tbh, I thought about it, but `RegisterCRDs` is also [used by](https://docs.cilium.io/en/v1.9/cmdref/cilium-agent/) cilium-agent which also provides `--skip-crd-creation`. Since I'm not an expert in the code, I decided to keep the common function untouched.
If you think it is reasonable to do so, I can implement it in another PR.

aanm replied:
@Fedosin The `RegisterCRDs` are not used by the cilium-agents anymore, that's why I suggested to also remove the flag option from it.

Fedosin replied:
@aanm alright, if you think it is safe, then I'll update my PR later today

kaworu commented on operator/flags.go:
```diff
@@ -315,5 +315,8 @@ func init() {
 	flags.String(option.BGPConfigPath, "/var/lib/cilium/bgp/config.yaml", "Path to file containing the BGP configuration")
 	option.BindEnv(option.BGPConfigPath)
 
+	flags.Bool(option.SkipCRDCreation, false, "Skip Kubernetes Custom Resource Definitions creations")
```
nit: I would reword the help message in a way that differ from what can be guessed by the flag name itself:
```suggestion
	flags.Bool(option.SkipCRDCreation, false, "when true, Kubernetes Custom Resource Definitions will not be created")
```

kaworu reviewed (changes requested):
Hi @Fedosin and thank you for the PR! [CI failure](https://github.com/cilium/cilium/pull/15655/checks?check_run_id=2324388109) is legit, please run `make -C install/kubernetes docs`.

kaworu reviewed (approved):
Helm related changes LGTM!

Fedosin committed:
Deprecate SkipCRDCreation option for cilium-agents

This commit deprecates option.Config.SkipCRDCreation and
--skip-crd-creation parameter for cilium-agents, as it's not
used there anymore.

Signed-off-by: Mike Fedosin <mfedosin@gmail.com>

christarazi commented on operator/flags.go:
```diff
@@ -315,5 +315,8 @@ func init() {
 	flags.String(option.BGPConfigPath, "/var/lib/cilium/bgp/config.yaml", "Path to file containing the BGP configuration")
 	option.BindEnv(option.BGPConfigPath)
 
+	flags.Bool(option.SkipCRDCreation, false, "when true, Kubernetes Custom Resource Definitions will not be created")
```
This description should match the previous flag

Fedosin replied:
@kaworu hi! should I revert your suggestion in this case? https://github.com/cilium/cilium/pull/15655#discussion_r611664659

christarazi replied:
Ah, I didn't see that Alex has already suggested this. Then I think keeping it is fine.

```suggestion
	flags.Bool(option.SkipCRDCreation, false, "When true, Kubernetes Custom Resource Definitions will not be created")
```

Fedosin replied:
done!

christarazi commented on operator/option/config.go:
```diff
@@ -41,6 +41,10 @@ const (
 	BGPConfigPath = "bgp-config-path"
 
+	// SkipCRDCreation specifies whether the CustomResourceDefinition will be
+	// disabled for the daemon
```
It's not for the daemon which refers to the cilium-agent though, so this should say Operator?

Fedosin replied:
agree, updated

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -880,6 +880,7 @@ func init() {
 	option.BindEnv(option.SelectiveRegeneration)
 
 	flags.Bool(option.SkipCRDCreation, false, "Skip Kubernetes Custom Resource Definitions creations")
+	flags.MarkDeprecated(option.SkipCRDCreation, "This flag is no longer available and will be removed in v1.12")
```
This will be deprecated in 1.10, and the flag will be removed in 1.11.

aanm reviewed (changes requested):
Also worth pointing out in the upgrade guide here https://github.com/cilium/cilium/blob/051032ce4e93121b94bc36cf1dfc8a51df0cc364/Documentation/operations/upgrade.rst#L401-L411

errordeveloper reviewed (changes requested):
This PR now needs rebase, as @aanm pointed out.

christarazi reviewed (dismissed):
Please squash your last commit into the commit that caused the change. No need for a separate commit
